### PR TITLE
added order entity

### PR DESCRIPTION
**Pull request by c-kalenyk**

### CODE REVIEW

File: src/main/java/mate/academy/bookstore/controller/OrderController.java [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+@RequestMapping("/orders")
+public class OrderController {
+    private final OrderService orderService;
+    private final OrderItemService orderItemService;
+    private final UserService userService;
+
+    @Operation(summary = "Get all orders from user",
+            description = "Get history of orders made by user")
+    @GetMapping
+    public Set<OrderDto> getOrdersHistory(Authentication authentication) {
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/controller/OrderController.java]:
```suggestion
    public Set<OrderDto> getOrdersHistory(Authentication authentication, @ParameterObject @PageableDefault Pageable pageable) {

```

File: src/main/java/mate/academy/bookstore/controller/OrderController.java [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+        UserResponseDto user = userService.findByEmail(authentication.getName());
+        return orderService.getUserOrders(user.getId());
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/controller/OrderController.java]:
```suggestion
        User user = (User) authentication.getPrincipal();
        return orderService. getUserOrders(user.getId(), pageable);
```

File: src/main/java/mate/academy/bookstore/controller/OrderController.java [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+    }
+
+    @Operation(summary = "Create new order", description = "Create new order from shopping cart")
+    @PostMapping
+    public OrderDto createOrder(Authentication authentication,
+                                @RequestBody @Valid CreateOrderRequestDto requestDto) {
+        UserResponseDto user = userService.findByEmail(authentication.getName());
+        return orderService.create(user.getId(), requestDto);
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/controller/OrderController.java]:
```suggestion
        User user = (User) authentication.getPrincipal();
        return orderService. create(user.getId(), requestDto);
```

File: src/main/java/mate/academy/bookstore/controller/OrderController.java [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+    @PreAuthorize("hasRole('ADMIN')")
+    @PatchMapping("/{id}")
+    public OrderDto updateOrderStatus(@PathVariable Long id,
+                                      @RequestBody @Valid UpdateOrderRequestDto requestDto) {
+        return orderService.updateOrderStatus(id, requestDto);
+    }
+
+    @Operation(summary = "Get order item", description = "Get order item by id")
+    @GetMapping("{orderId}/items/{itemId}")
+    OrderItemDto getOrderItem(@PathVariable Long orderId,
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/controller/OrderController.java]:
```suggestion
    public OrderItemDto getOrderItem(@PathVariable Long orderId,
```

File: src/main/java/mate/academy/bookstore/controller/OrderController.java [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+                              @PathVariable Long itemId) {
+        return orderItemService.getByOrderIdAndItemId(orderId, itemId);
+    }
+
+    @Operation(summary = "Get all items for a specific order")
+    @GetMapping("/{orderId}/items")
+    Set<OrderItemDto> getAllOrderItems(@PathVariable Long orderId) {
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/controller/OrderController.java]:
```suggestion
    public Set<OrderItemDto> getAllOrderItems(@PathVariable Long orderId, Authentication authentication) {
```

File: src/main/java/mate/academy/bookstore/controller/OrderController.java [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+    @PatchMapping("/{id}")
+    public OrderDto updateOrderStatus(@PathVariable Long id,
+                                      @RequestBody @Valid UpdateOrderRequestDto requestDto) {
+        return orderService.updateOrderStatus(id, requestDto);
+    }
+
+    @Operation(summary = "Get order item", description = "Get order item by id")
+    @GetMapping("{orderId}/items/{itemId}")
+    OrderItemDto getOrderItem(@PathVariable Long orderId,
+                              @PathVariable Long itemId) {
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/controller/OrderController.java]:
```suggestion
                              @PathVariable Long itemId, Authentication authentication) {
```
get user Id and provide it to service method
same for getAllOrderItems

File: src/main/java/mate/academy/bookstore/model/Order.java [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+@Getter
+@Setter
+@SQLDelete(sql = "UPDATE orders SET is_deleted = TRUE WHERE id = ?")
+@SQLRestriction("is_deleted = FALSE")
+@Table(name = "orders")
+public class Order {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @ManyToOne(fetch = FetchType.LAZY)
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/model/Order.java]:
```suggestion
    @ManyToOne(fetch = FetchType.LAZY, optional = false)
```

File: src/main/java/mate/academy/bookstore/model/Order.java [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+    @JoinColumn(name = "user_id", nullable = false)
+    private User user;
+    @Column(nullable = false)
+    @Enumerated(EnumType.STRING)
+    private Status status;
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/model/Order.java]:
```suggestion
    private Status status = Status.PENDING;
```

File: src/main/java/mate/academy/bookstore/model/Order.java [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+    @Column(nullable = false)
+    private BigDecimal total;
+    @Column(nullable = false)
+    private LocalDateTime orderDate;
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/model/Order.java]:
```suggestion
    private LocalDateTime orderDate = LocalDateTime.now();
```

File: src/main/java/mate/academy/bookstore/model/Order.java [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+    @Column(nullable = false)
+    private String shippingAddress;
+    @OneToMany(mappedBy = "order", fetch = FetchType.LAZY)
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/model/Order.java]:
```suggestion
    @OneToMany(mappedBy = "order", cascade = CascadeType.ALL, orphanRemoval = true)
```

File: src/main/java/mate/academy/bookstore/model/OrderItem.java [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+import lombok.Setter;
+
+@Entity
+@Getter
+@Setter
+@Table(name = "order_items")
+public class OrderItem {
+    @Id
+    private Long id;
+    @ManyToOne(fetch = FetchType.LAZY)
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/model/OrderItem.java]:
```suggestion
    @ManyToOne(fetch = FetchType.LAZY, optional = false)
```

File: src/main/java/mate/academy/bookstore/model/OrderItem.java [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+    @JoinColumn(name = "order_id", nullable = false)
+    private Order order;
+    @OneToOne(fetch = FetchType.LAZY)
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/model/OrderItem.java]:
```suggestion
    @OneToOne(fetch = FetchType.LAZY, optional = false)
```

File: src/main/java/mate/academy/bookstore/service/impl/OrderItemServiceImpl.java [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+import mate.academy.bookstore.mapper.OrderItemMapper;
+import mate.academy.bookstore.model.OrderItem;
+import mate.academy.bookstore.repository.orderitem.OrderItemRepository;
+import mate.academy.bookstore.service.OrderItemService;
+import org.springframework.stereotype.Service;
+
+@Service
+@RequiredArgsConstructor
+public class OrderItemServiceImpl implements OrderItemService {
+    private final OrderItemRepository orderItemRepository;
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/service/impl/OrderItemServiceImpl.java]: replace all logic to order service

File: src/main/java/mate/academy/bookstore/service/impl/OrderServiceImpl.java [hunk 14]
```diff
@@ -0,0 +1,88 @@
...
+    private final UserRepository userRepository;
+    private final CartItemRepository cartItemRepository;
+    private final OrderItemRepository orderItemRepository;
+
+    @Override
+    public OrderDto create(Long userId, CreateOrderRequestDto requestDto) {
+        Set<CartItem> cartItems = cartItemRepository.getCartItemsByShoppingCartId(userId);
+        if (cartItems.isEmpty()) {
+            throw new EntityNotFoundException("Can't create empty order");
+        }
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/service/impl/OrderServiceImpl.java]:
```suggestion
        ShoppingCart cart = cartRepository.findByUserId(userId);
        if (cart.getCartItems().isEmpty()) {
            throw new OrderProcessingException("Cart is empty for user: " + userId);
        }
```

File: src/main/java/mate/academy/bookstore/service/impl/OrderServiceImpl.java [hunk 15]
```diff
@@ -0,0 +1,88 @@
...
+                new EntityNotFoundException("Can't find user by id: " + userId));
+        order.setUser(user);
+        order.setStatus(Order.Status.PENDING);
+        order.setTotal(calculateTotalPrice(orderItems));
+        order.setOrderDate(LocalDateTime.now());
+        order.setShippingAddress(requestDto.getShippingAddress());
+        order.setOrderItems(orderItems);
+        orderRepository.save(order);
+        orderItemRepository.saveAll(orderItems);
+        return orderMapper.toDto(order);
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/service/impl/OrderServiceImpl.java]:
```suggestion
        Order order = orderMapper.cartToOrder(cart, orderDto.shippingAddress());
        cart.clearCart();
        return orderMapper.toOrderDto(orderRepository.save(order));
```

where in OrderMapper
```
@Mapping(target = "id", ignore = true)
@Mapping(target = "total", source = "cart.cartItems", qualifiedByName = "total")
@Mapping(target = "orderItems", source = "cart.cartItems")
Order cartToOrder(ShoppingCart cart, String shippingAddress);
```
and 
```
@AfterMapping
default void updateOrder(@MappingTarget Order order) {
    order.getOrderItems().forEach(oi -> oi.setOrder(order));
}

@Named("total")
default BigDecimal getTotal(Set<CartItem> cartItems) {
    return cartItems.stream()
                .map(i -> i.getBook().getPrice().multiply(BigDecimal.valueOf(i.getQuantity())))
                .reduce(BigDecimal.ZERO, BigDecimal::add);
}
```

And in OrderItemMapper

```
@Mapping(target = "id", ignore = true)
@Mapping(target = "price", source = "book.price")
OrderItem cartItemToOrderItem(CartItem cartItem);
```

And in OrderMapper
```
@Mapping(target = "orderDate", dateFormat = "yyyy-MM-dd HH:mm")
@Mapping(target = "userId", source = "user.id")
OrderDto toOrderDto(Order order);
```

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

**Review by MentorMateAcademy — APPROVED** [review]

**Pull request merged**